strawberryfields/backends/states.py

```diff
@@ -1405,3 +1415,412 @@ def all_fock_probs(self, **kwargs):
         mu = self._mu
         cov = self._cov
         return twq.probabilities(mu, cov, cutoff, hbar=self._hbar)
+
+
+class BaseBosonicState(BaseState):
+    r"""Class for the representation of quantum states as superpositions of
+    Gaussian states.
+    """
+    # pylint: disable=too-many-public-methods
+
+    def __init__(self, state_data, num_modes, num_weights, mode_names=None):
+        super().__init__(num_modes, mode_names)
+
+        self._data = state_data
+        self.num_weights = num_weights
+        # vector of means and covariance matrix, using frontend x,p scaling
+        self._mus = self._data[0] * np.sqrt(self._hbar / 2)
+        self._covs = self._data[1] * (self._hbar / 2)
+        self._weights = self._data[2]
+
+        self._basis = "bosonic"
+        self._str = "<BosonicState: num_modes={}, num_weights={}, pure={}, hbar={}>".format(
+            self.num_modes, self.num_weights, self._pure, self._hbar
+        )
+
+    def __eq__(self, other):
+        """Equality operator for BaseBosonicState.
+
+        Returns True if other BaseBosonicState is close to self.
+        This is done by comparing the weights, means vectors and covs matrices.
+        If both are within the EQ_TOLERANCE, True is returned.
+
+        Args:
+            other (BaseGaussianState): BaseGaussianState to compare against.
+        """
+        # pylint: disable=protected-access
+        # TODO: check equality for two different representations of the same state.
+        # This only checks if they have equal representations.
+        if not isinstance(other, type(self)):
+            return False
+
+        if self.num_modes != other.num_modes:
+            return False
+
+        if self.num_weights != other.num_weights:
+            return False
+
+        if (
+            np.allclose(self._mus, other._mus, atol=self.EQ_TOLERANCE, rtol=0)
+            and np.allclose(self._covs, other._covs, atol=self.EQ_TOLERANCE, rtol=0)
+            and np.allclose(self._weights, other._weights, atol=self.EQ_TOLERANCE, rtol=0)
+        ):
+            return True
+
+        return False
+
+    def means(self):
+        r"""The vectors of means describing the Bosonic state.
+
+        Returns:
+          array: a num_weights by :math:`2N` array.
```

Comment:
```suggestion
          array[float]: an array of shape ``(num_weights, 2N)``
```
Also, ``N`` is not defined here?

Reply:
```suggestion
          array[complex]: an array of shape ``(num_weights, 2N)``
```